grader_utils/__init__.py

```diff
@@ -0,0 +1,196 @@
+        :param file stdin: Same as stdout but for stdin. Note that input and delayed_inputs
+            will only work if this is set to PIPE.
+        :param str cwd: Set the directory for the process.
+        :param str input: String to feed in for stdin. Only works if stdin is set to PIPE.
+            Set to None to not feed in anything. Default is None.
+        :param list[str] | list[int] delayed_inputs: Will iterate through list and for
+            each entry input the string into
+            stdin or wait the given amount of time. Default is [].
+        :param bool check_pg_alive: Indicate whether to use our is_alive() or
+            Thread.is_alive() in the join() function. Default is True since we
+            want to check for running child processes after the main process
+            exits.
+        """
+
+        super(ProcessWrapper, self).__init__()
+        self.command = (shlex.split(command) if type(command) is str
+                        else command)
+        self.timeout = kwargs.get('timeout', None)
+        self.stdout = kwargs.get('stdout', DEVNULL)
+        self.stderr = kwargs.get('stderr', DEVNULL)
+        self.stdin = kwargs.get('stdin', DEVNULL)
+        self.cwd = kwargs.get('cwd', None)
+        self.inputs = kwargs.get('input', None)
+        self.delayed_inputs = kwargs.get('delayed_inputs', [])
+        self.disable_communicate = kwargs.get('disable_communicate', False)
+        self.check_pg_alive = kwargs.get('check_pg_alive', True)
+        self.env = kwargs.get('env', None)
+
+        # Keep track of subprocess and thread that waits on it
+        self.process = None
+        self.thread = None
+
+        # Fields that will be outputted later
+        self.killed = False
+        self.stdout_res = None
+        self.stderr_res = None
+        # If the process timeouts
+        self.exit_timeout = False
+        # If the process is killed by SIGKILL or exit normally
+        self.exit_force = False
+        # If
+        # - the process timeouts,
+        # - or any of the processes in the process group timeouts while main
+        #   process exited.
+        # Either case we will try to kill the process group using SIGTERM.
+        self.exit_timeout_pg = False
+        # If
+        # - the process is not killed by SIGTERM
+        # - or any of the processes in the process group is not killed by
+        #   SIGTERM.
+        # In either case we will try to kill the process group using SIGKILL.
+        self.exit_force_pg = False
+
+    def run(self):
+        """Run the command as a separate process and return immediately.
+
+        Call join() later to get result of process. Timeout does not start
+        until join() is called. Additionally send the delayed_inputs and inputs
+        to process.
+        """
+        def target():
+            # Pass these inputs to STDIN with delays
+            for i in self.delayed_inputs:
+                if type(i) is int or type(i) is float:
```

Comment:
I don't like this. Can we instead define a type like `Sleep` or something that stores a float and any other type gets cast to bytes or is rejected?